docs/proposals/fluentd-limit-requests-to-apiserver.md

```diff
@@ -0,0 +1,107 @@
+
+The pod metadata is currently sourced by watching the Openshift API server for Pod information via the [Kubernetes Metadata Filter](https://github.com/fabric8io/fluent-plugin-kubernetes_metadata_filter)[2]; Each watch maintains an open connection to the API server. Testing at larger scales of Openshift clusters has shown issues where Fluentd is unable to connect to the API server because it is limited by the **maxRequestsInFlight** configuration setting [1].  The current deployment topology at scale:
+
+* Results in logs not being shipped to the aggregator and thus being unavailable
+* Limits the number of cluster nodes because of the cap on the number of open connections
+
+This proposal addresses both the short and long term solutions to resolve the connection issues being experienced using the v3.4 EFK stack.
+### Considerations
+1. OpenShift Logging is not the only consumer of OpenShift log messages.  There are customers who desire to use their existing logging infrastructure(e.g. Splunk, Elasticsearch).  Log entries need to be enriched with metadata irrespective of the final destination.
+2. It is speculated that there is a low probability that labels and annotations will change during the lifetime of a running pod. We should determine if 'stale' labels and annotations are acceptable
+3. Identify the tolerance for dropped message.
+
+## Proposal - Short Term (3.4)
+1. Leave the value as-is and note that it must be adjusted to account for # of nodes plus some headroom (swag @ 10%?).
+1. Increase the default to our current maximum supported node count, plus some headroom.
+1. Increasing maxRequestsInflight upstream
+1. **COMPLETE** ~~Modify Fluentd Metadata plugin to cache by namespace and pod name only~~[2][4]
+
+## Proposal - Long Term
+### Option 1 - Admission Controller
+1. Create an admission controller which embeds the necessary metadata as envirnment variables into the pods on submission.  In this way every pod that is run on the system will have its data locally.
+1. Mount the docker socket into the Fluentd pod such that it can query container's environment variables.
+1. Remove any watches from the Fluentd pods to eliminating excessive traffic from the node.
+
+#### Limitations
+1. Labels and annotations will never be updated since docker containers are immutable.
+1. It binds the log normalization to the container runtime.
+
+This will require:
+* Admission controller introduced to [Openshift Origin](https://github.com/openshift/origin)
```

Comment:
Needs to be a Kubernetes solution, not just OpenShift, and as mentioned I don't think an AC will work for this because it pollutes each pod.  Instead, we need to be able to get this data efficiently on the node from the container runtime or kubelet.